sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,252 @@
+        for index, col in enumerate(X.T):
+            if index not in self.continuous_columns:
+                continue
+            intervals = self._mdlp(col, Y, min_depth)
+            intervals.sort(key=lambda (a, b): a)
+            self.intervals[index] = [(interval, i)
+                                     for i, interval in enumerate(intervals)]
+
+    def transform(self, X):
+        """Converts the continuous values in X into ascii character
+        values. The mapping is defined by self.intervals.
+        """
+        assert self.continuous_columns is not None, "You must fit the object " \
+                                                    "before transforming data."
+        discretized = list()
+        for index, col in enumerate(X.T):
+            if index not in self.continuous_columns:
+                discretized.append(col.reshape(len(X), 1))
+            transformed = self.cts2cat(index, col)
+            discretized.append(transformed.reshape(len(X), 1))
+        return np.hstack(discretized)
+
+    def fit_transform(self, X, Y, continuous_columns=[], min_depth=None):
+        """Performs the fitting and the data transformations.
+        """
+        self.fit(X, Y, continuous_columns, min_depth)
+        return self.transform(X)
+
+    def cat2intervals(self, index, col):
+        """Converts a categorical column into an interval. This is the naive
+        linear search method, but in practice, because there won't be many
+        intervals to search through (usually at most 5), this works effectively
+        without an interval tree.
+        """
+        mappings = self.intervals[index]
+        out_intervals = list()
+        for cat in col:
+            for interval, currcat in mappings:
+                if cat == currcat:
+                    out_intervals.append(interval)
+        return out_intervals
+
+    def bin(self, index, attr):
+        """Bins an attribute into its appropriate interval.
+        """
+        mappings = self.intervals[index]
+        for (a, b), _ in mappings:
+            if a < attr <= b: return (a, b)
+        raise ValueError("Numeric value did not fit in any interval")
+
+    def cts2cat(self, index, col):
+        """Converts each continuous value into a categorical value.
+        """
+        mappings = self.intervals[index]
+        categorical = list()
+        for attr in col:
+            for (a, b), cat in mappings:
+                if a < attr <= b:
+                    categorical.append(cat)
+                    break
+        assert len(categorical) == len(col)
+        return np.array(categorical)
+
+    def _mdlp(self, attributes, Y, min_depth):
+        """
+        *attributes*: A python list of continuous values.
+
+        *Y*: A python list of 0's and 1's
+
+        Returns a list of tuples representing intervals.
+        """
+
+        # attr_list contains triples, (attribute, y class)
+        attr_list = list([attr, y] for attr, y in izip(attributes, Y))
```

Comment:
Instead, maintain parallel numpy arrays for attr (let's call it `x`, a column of `X`) and `y`.

```python
order = np.argsort(x)
x = x[order]
y = y[order]
```
Rewrite the method to work with this data structure.